### PR TITLE
feat: link oauth accounts with same email

### PR DESCRIPTION
**Pull request by ZeRego**
Closes: #9692 

### Description:

Note: I'm adding unit tests to cover all the `loginWithOpenId` scenarios in the following PR.

https://github.com/lightdash/lightdash/assets/9117144/560d633e-760e-444f-b23d-30ea4b46fcf9

### Reviewer actions

- [ ] I have manually tested the changes in the preview environment
- [ ] I have reviewed the code
- [ ] I understand that "request changes" will block this PR from merging

### CODE REVIEW

File: packages/backend/src/services/UserService.ts [hunk 1]
```diff
@@ -518,7 +520,28 @@ export class UserService extends BaseService {
...
+                await this.openIdIdentityModel.findIdentitiesByEmail(
+                    openIdUser.openId.email,
+                );
+            const identitiesUsers = uniq(
+                identities.map((identity) => identity.userUuid),
+            );
+            if (identitiesUsers.length > 1) {
+                Logger.warn(
+                    `Multiple openid identities found with the same email ${openIdUser.openId.email}`,
+                );
```
Comment (ZeRego) [inline packages/backend/src/services/UserService.ts:536]: Warning to cover edge cases with multiple authentications providers enabled.

File: packages/backend/src/models/OpenIdIdentitiesModel.ts [hunk 2]
```diff
@@ -53,6 +53,14 @@ export class OpenIdIdentityModel {
             );
     }
 
+    async findIdentitiesByEmail(email: string): Promise<OpenIdIdentity[]> {
```
Comment (owlas) [inline packages/backend/src/models/OpenIdIdentitiesModel.ts:56]: this returns all OpenId's with an openid email matching the query email and doesn't account for primary emails on Lightdash user accounts ✅

File: packages/backend/src/services/UserService.ts (hunk 1 above)
Comment (owlas) [reply packages/backend/src/services/UserService.ts:536]: great idea!

**Review by owlas — APPROVED** [review]

File: packages/backend/src/services/UserService.ts [hunk 3]
```diff
@@ -518,7 +520,28 @@ export class UserService extends BaseService {
...
+            } else if (identitiesUsers.length === 1) {
+                sessionUser = await this.userModel.findSessionUserByUUID(
+                    identitiesUsers[0],
+                );
+            }
+        }
+
+        // Link openid identity to existing user
```
[the remark this answers is not in the document]
Comment (ZeRego) [reply packages/backend/src/services/UserService.ts:544]: Exported logic that links openid with a user to a private method so we don't mix the two scenarios here.

**Pull request merged**